### PR TITLE
SPM Performance and error correction

### PR DESCRIPTION
**Pull request by eanders**
…ged; catch missing client

[//]: # 'remove this if PR is for a release-* branch'
# _Please squash merge this PR_

## Description

This changes the SPM to only load the services needed for the report.  Additionally it attempts to catch more situations where source data has disappeared.

This also fixes the following bugs:
* The SPM was still including projects from all types if anything other than the project type was chosen in the filters when running the SPM
* The HDX upload was being generated with an incorrect `report_type` so you couldn't drill into the detail tables
* The HDX upload wasn't correctly limiting the projects included to those in the SPM unless the SPM was run only choosing project types 
* Service preloading was extremely slow when clients have long NbN enrollments, this moves those to a pluck that only brings the necessary columns, and only returns services within the report range
* Enrollments for the SPM do not need the entire width of the Enrollment table, to save memory and time, we now pull a subset of columns when loading Enrollments.
* In some situations, the HDX upload was attempting to run for data sets with no data multiple times.  If we don't find any projects in the project type, these are now skipped and not retried.
* The SPM enrollment filter was using project types codes where it needed numbers

## Type of change
[//]: # 'remove options that are not relevant'
- [x] Bug fix

## Checklist before requesting review
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] My code includes comments and/or descriptive variable names to help other engineers understand the intent (or not applicable)
- [x] My code follows the style guidelines of this project (rubocop)
- [x] I have updated the documentation (or not applicable)
- [x] If it's not obvious how to test this change, I have provided testing instructions in this PR or the related issue

### CODE REVIEW

**Review by dtgreiner — APPROVED** [review]

File: drivers/hud_apr/app/controllers/hud_apr/dq/cells_controller.rb [hunk 1]
```diff
@@ -7,6 +7,7 @@
 module HudApr::Dq
   class CellsController < HudApr::CellsController
     include DqConcern
+    before_action :generator
```
Comment (ttoomey) [inline drivers/hud_apr/app/controllers/hud_apr/dq/cells_controller.rb:11]: Why do we call this as a before_action and not lazy load? Maybe add comment if there a side-effect we need?

File: drivers/hud_spm_report/spec/models/fy2023/active_record_preload_spec.rb [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+  describe 'preloads work as expected' do
+    it '4 services are created' do
+      expect(GrdaWarehouse::Hud::Service.count).to eq(4)
+    end
+    it '2 services are between 2022-01-02 and 2022-01-15' do
+      expect(GrdaWarehouse::Hud::Service.where(DateProvided: '2022-01-02'.to_date .. '2022-01-15'.to_date).count).to eq(2)
+    end
+    it 'when preloading, one enrollment has 4 services' do
+      expect(GrdaWarehouse::Hud::Enrollment.preload(:services).first.services.count).to eq(4)
```
Comment (ttoomey) [inline drivers/hud_spm_report/spec/models/fy2023/active_record_preload_spec.rb]: Is the intention of these assertions to test if it preloaded 4 services? If so should this use size instead of count? Size will use the preloaded records if available while count ignores preloaded records and makes a new query

File: drivers/hud_spm_report/app/models/hud_spm_report/fy2023/episode.rb [hunk 3]
```diff
@@ -148,15 +148,17 @@ def compute_episode(enrollments, included_project_types:, excluded_project_types
...
-              map do |service|
-                date = service.date_provided
-                next unless service.record_type == HudUtility2024.record_type('Bed Night', true) && date.between?(first_night, last_night)
-
-                [enrollment, service.id, date]
+            enrollment_services.
+              map.with_index do |date, i|
+                next unless date.between?(first_night, last_night)
+
+                [enrollment, i, date]
```
Comment (ttoomey) [inline drivers/hud_spm_report/app/models/hud_spm_report/fy2023/episode.rb:163]: small scope creep but maybe candidate_bed_nights could use a little documentation about the return value. In particular, what's the intention of the index `i`

File: drivers/hud_spm_report/app/models/hud_spm_report/generators/fy2023/hdx_upload.rb [hunk 4]
```diff
@@ -213,29 +213,49 @@ def spm(table_name, cell_name)
     def dq(section, table_name, cell_name)
       return unless RailsDrivers.loaded.include?(:hud_apr)
 
+      @attempted ||= Set.new
+      @reports ||= {}
+
+      # prevent retrying reports that don't have any projects
+      return '' if @attempted.include?(section) && @reports[section].nil?
```
Comment (ttoomey) [inline drivers/hud_spm_report/app/models/hud_spm_report/generators/fy2023/hdx_upload.rb:221]: do we use an empty string because this is written to a non-nullable col?

File: drivers/hud_spm_report/app/models/hud_spm_report/fy2023/episode_batch.rb [hunk 5]
```diff
@@ -49,16 +80,17 @@ def calculate_batch(client_ids)
     end
 
     # The associations seem to make imports run one at a time, so, they are passed separately in parallel arrays
-    private def save_episodes!(episodes, bed_nights, enrollment_links)
+    private def save_episodes!(episodes, _bed_nights, enrollment_links)
```
Comment (ttoomey) [inline drivers/hud_spm_report/app/models/hud_spm_report/fy2023/episode_batch.rb:85]: We're leaving the bed_nights var because we might want to enable this in the future?

**Review by ttoomey — APPROVED** [review]
Looks good, thanks for the performance improvement!

File: drivers/hud_apr/app/controllers/hud_apr/dq/cells_controller.rb (hunk 1 above)
Comment (eanders) [reply drivers/hud_apr/app/controllers/hud_apr/dq/cells_controller.rb:11]: No particular reason, just following the existing pattern.  I _think_ `generator` isn't tagged as `helper_method` but I may have missed that.  We probably call `@generator` and this was failing prior to adding this.

Definitely worth revisiting in the future.

File: drivers/hud_spm_report/app/models/hud_spm_report/fy2023/episode.rb (hunk 3 above)
Comment (eanders) [reply drivers/hud_spm_report/app/models/hud_spm_report/fy2023/episode.rb:163]: Definitely.  I don't like this pattern, returning a triple, but wanted to remain compatible with the current API.  The middle item was the services ID, and I believe just needs to be unique within the enrollment.

File: drivers/hud_spm_report/app/models/hud_spm_report/fy2023/episode_batch.rb (hunk 5 above)
Comment (eanders) [reply drivers/hud_spm_report/app/models/hud_spm_report/fy2023/episode_batch.rb:85]: Yes, we disabled it to prevent saving hundreds of thousands of rows that aren't currently displayed in the UI, but wanted to leave the pattern behind in case we want to re-enable that in the future.
